File: src/0560-Subarray-Sum-Equals-K/0560.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using  namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    int subarraySum(vector<int>& nums, int k) 
    {
        int result = 0, cur_sum = 0;
        unordered_map<int, int> sum_dict = {{0, 1}};
        for (auto& num : nums)
        {
            cur_sum += num;
            result += sum_dict[cur_sum - k];
            sum_dict[cur_sum]++;
        }
        return result;
    }
};
int main()
{
    vector<int> nums = {1, 1, 1, 1, 1};
    int k = 3;
    cout << Solution().subarraySum(nums, k);
    return 0;
}

File: src/0560-Subarray-Sum-Equals-K/0560.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        res, cur = 0, 0
        sum_dict = collections.defaultdict(int)
        sum_dict[0] = 1
        
        for num in nums:
            cur += num
            res += sum_dict[cur - k]
            sum_dict[cur] += 1
        return res

if __name__ == "__main__":
    nums = [1,1,1]
    k = 2
    print(Solution().subarraySum(nums, k))